File: Work In Progress/scripts/objective-c/ShutNet.m
#import <Foundation/Foundation.h>
#import <NMSSH/NMSSH.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSMutableArray *ipAddresses = [[NSMutableArray alloc] init];

        for (int i = 1; i <= 254; i++) {
            NSString *ip = [NSString stringWithFormat:@"192.168.1.%d", i];
            NSLog(@"Scanning %@", ip);
            NSTask *task = [[NSTask alloc] init];
            task.launchPath = @"/sbin/ping";
            task.arguments = @[@"-c", @"1", @"-t", @"1", ip];
            NSPipe *pipe = [[NSPipe alloc] init];
            task.standardOutput = pipe;
            [task launch];
            [task waitUntilExit];
            if ([task terminationStatus] == 0) {
                [ipAddresses addObject:ip];
            }
        }
        
        NSURL *fileURL = [NSURL fileURLWithPath:@"Ips.txt"];
        NSString *ipString = [ipAddresses componentsJoinedByString:@"\n"];
        [ipString writeToURL:fileURL atomically:YES encoding:NSUTF8StringEncoding error:nil];

        for (NSString *ip in ipAddresses) {
            NSLog(@"Shutting down %@", ip);
            NSError *error = nil;
            NMSSHSession *session = [NMSSHSession connectToHost:ip withUsername:@"username"];
            [session authenticateByPassword:@"password"];
            if (session.isAuthorized) {
                NSString *computerName = [session.channel execute:@"hostname" error:&error];
                if (error) {
                    NSLog(@"Could not resolve computer name for IP address %@", ip);
                    continue;
                }
                NSString *osName = [session.channel execute:@"uname" error:&error];
                if (error) {
                    NSLog(@"Could not determine operating system for %@", computerName);
                    continue;
                }
                if ([osName containsString:@"Windows"]) {
                    [session.channel execute:@"shutdown.exe /s /t 0" error:nil];
                } else if ([osName containsString:@"Linux"]) {
                    [session.channel execute:@"sudo shutdown -h now" error:nil];
                } else if ([osName containsString:@"Darwin"]) {
                    [session.channel execute:@"sudo shutdown -h now" error:nil];
                } else {
                    NSLog(@"Unknown operating system for %@", computerName);
                }
            } else {
                NSLog(@"Could not authenticate with %@", ip);
            }
            [session disconnect];
        }

        [[NSFileManager defaultManager] removeItemAtURL:fileURL error:nil];
    }
    return 0;
}

File: Work In Progress/scripts/swift/ShutNet.swift
import Foundation
import SwiftSSH

var ipAddresses: [String] = []

for i in 1...254 {
    let ip = "192.168.1.\(i)"
    print("Scanning \(ip)")
    let task = Process()
    task.launchPath = "/sbin/ping"
    task.arguments = ["-c", "1", "-t", "1", ip]
    let pipe = Pipe()
    task.standardOutput = pipe
    task.launch()
    task.waitUntilExit()
    if task.terminationStatus == 0 {
        ipAddresses.append(ip)
    }
}

let fileURL = URL(fileURLWithPath: "Ips.txt")
let ipString = ipAddresses.joined(separator: "\n")
try! ipString.write(to: fileURL, atomically: true, encoding: .utf8)

for ip in ipAddresses {
    print("Shutting down \(ip)")
    guard let computerName = try? SSH(host: ip, username: "username", password: "password").capture("hostname") else {
        print("Could not resolve computer name for IP address \(ip)")
        continue
    }
    let osName = try? SSH(host: ip, username: "username", password: "password").capture("uname")
    if osName?.contains("Windows") == true {
        try? SSH(host: ip, username: "username", password: "password").capture("shutdown.exe /s /t 0")
    } else if osName?.contains("Linux") == true {
        try? SSH(host: ip, username: "username", password: "password").capture("sudo shutdown -h now")
    } else if osName?.contains("Darwin") == true {
        try? SSH(host: ip, username: "username", password: "password").capture("sudo shutdown -h now")
    } else {
        print("Unknown operating system for \(ip)")
    }
}

// Clear the IP addresses file
try? FileManager.default.removeItem(at: fileURL)
